Fix streaming request bodies

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

FIX #26904

At this point I'm not sure if the hyper `Body::wrap_stream` integration by way of the IPC route is broken, or if perhaps there is a problem with the WPT server not handling the requests properly, as I am getting a bunch of:

```
 0:04.54 INFO STDERR: 127.0.0.1 - - [13/Jun/2020 18:22:08] code 400, message Bad request syntax ('4')
 0:04.54 INFO STDERR: 127.0.0.1 - - [13/Jun/2020 18:22:08] "4" 400 -
```

In any case, the `start_reading` call in `body.rs` had to be put into the other function, since it should not be called for each chunk(this was actually failing before). 

I've also added the chunked header in case the source is a stream.

Last thing to figure out, why is the request data not actually reaching the server, or not being handled properly by it. 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

gterzian committed:
fix streaming request bodies, terminate fetch if the body stream errors

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/body.rs
 * @kichjang: components/script/body.rs, components/net/http_loader.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify net and script code, but no tests are modified. Please consider adding a test!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26810) made this pull request unmergeable. Please resolve the merge conflicts.

gterzian commented:
So I have:

1. Removed the use of `Body::wrap_stream`, because it wasn't working. I'm not sure if this was a client or server-side issue, see also https://github.com/web-platform-tests/wpt/issues/24139
2. Ensure we call `start_reading` on the stream only when the route is first connected to.
3. added a mechanism to propagate read errors in script to a network error in net. 

So this makes the relevant tests pass.

@Manishearth r?

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 0dc231b40892a41c3b58ba0241639c3a769f90ea with merge ef22347a0dfdb8f0801cd036b8934fadbc67db1b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/f3Mgih6KRKKnCMYaxbDLGA)

gterzian commented:
Ok so the problem is now that I've switched the `start_reading` operation to read on the stream, this happens multiple times in case of a re-direct, so it fails. 

However, I just need to add a conditional to this based on the source, so that it only happens when we're using an actual `ReadableStream`, and by definition in these cases there will never be a re-direct.

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 45e56cb514db330aff71fd43703ebfdf9aa82240 with merge 8b9af9994408ea5ee71eb9863c5eb4733b1b9edf...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/DLWwW680T1OxKd98sKPODg)
State: approved= try=True

gterzian commented:
Actually I think I'll keep the chunking for when it's a stream body, what we need to do is fix the WPT server, see https://github.com/web-platform-tests/wpt/issues/24139

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 72d09f07fd2606c1691c91013b416830929d46ed with merge 187e02a11227cb939d1c5631a18f65ac3b33ac56...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/dO35sOpRRKWoAzWrcdZv5w)

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 607dc94ef617b86fceb21b8ffcba00689ebf0a60 with merge 38df335f54b533ed72341f598c02b549a0982219...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/X-IBqBRwSe6EAAz7eOFqRA)

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 719b395c403d2c409b19ecfbb334219afcb205ad with merge 07e2e5918ed2d052a6836aa8f0b2fd38a221c04c...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/VtW42rWbS8CJaRhDK1JM6w)

gterzian commented:
Failure looks like https://github.com/servo/servo/issues/23290

gterzian commented:
@Manishearth r?

So the one test is still failing, because the WPT server doesn't support chunked requests(https://github.com/web-platform-tests/wpt/issues/24139), also this isn't a feature actually implemented anywhere yet. 

Re the non-stream case, I've added some explicit buffering so that it matches the previous behavior when the body of a request was a `Vec<u8>`, note that I think hyper could be relied on to do this anyway, however I'd rather be sure and explicit.

Manishearth commented:
@bors-servo r+

this looks great, thanks!

bors-servo commented:
:pushpin: Commit 719b395 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 719b395c403d2c409b19ecfbb334219afcb205ad -->

bors-servo commented:
:hourglass: Testing commit 719b395c403d2c409b19ecfbb334219afcb205ad with merge 16a99daa229dec77ac280c82216936b97c8e4d9c...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/DGwW2xY5S9a43lsKhhqDyQ)
Approved by: Manishearth
Pushing 16a99daa229dec77ac280c82216936b97c8e4d9c to master...
